### PR TITLE
GFORMS-961 Support for optionally hiding page titles in add-to-list via presentationHint

### PR DESCRIPTION
**Pull request by girishkamat**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala [hunk 1]
```diff
@@ -46,11 +48,34 @@ object Section {
...
+        (JsPath \ "includeIf").readNullable[IncludeIf] and
+        (JsPath \ "repeatsMax").readNullable[Expr] and
+        (JsPath \ "pages").read[NonEmptyList[Page]] and
+        (JsPath \ "addAnotherQuestion").read[FormComponent] and
+        (JsPath \ "instruction").readNullable[Instruction] and
+        (JsPath \ "presentationHint").read[Option[List[PresentationHint]]](
+          (json: JsValue) =>
+            parseOpt(JsDefined(json), PresentationHintParser.validate)
+              .fold(us => JsError(us.toString), fv => JsSuccess(fv)))
+    )(AddToList.apply _)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala:72]: I don't like this explicit `Reads`. Cannot we just define `Reads[PresentationHint]` which will do the parsing?

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala (hunk 1 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala]: Thanks @josef-vlach. I have replaced this with a `Reads[List[PresentationHint]]`. Have made it a list to mimic what is done on FormComponent level

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala [hunk 2]
```diff
@@ -46,11 +47,21 @@ object Section {
...
 
+  object AddToList {
+
+    implicit val presentationHintReads: Reads[List[PresentationHint]] = (json: JsValue) =>
+      parseOpt(JsDefined(json), PresentationHintParser.validate)
+        .fold(us => JsError(us.toString), fv => JsSuccess(fv.getOrElse(List.empty)))
+
+    implicit val format: Format[AddToList] = Json.format[AddToList]
+  }
+
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala:64]: I was hoping in more subtle solution, like hiding this `Reads[List[PresentationHint]]` in `def readAddToList(json: JsValue)` and not implementing it in terms of `parseOpt`.

Something like:

```
private def readAddToList(json: JsValue) = {
  implicit def presentationHintsReads: OFormat[List[PresentationHint]] =
    OFormatWithTemplateReadFallback {
      case JsString(str) => PresentationHintParser.validate(str).fold(e => JsError(e.error), JsSuccess(_))
      case unknown       => JsError("Expected String, got " + unknown)
    }
  Json.format[Section.AddToList].reads(json)
}
```

Then we can revert `toOpt` changes etc.

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala (hunk 2 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala:64]: This is still a dirty solution, though. The best thing would be to introduce newtype for `List[PresentationHint]` and do the logic there.

Comment (girishkamat) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Section.scala]: Implemented as per the discussion

**Review by josef-vlach — APPROVED** [review]

**Pull request merged**